src/app-layout/__tests__/dynamic-overlap.test.tsx

```diff
@@ -0,0 +1,94 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import React from 'react';
+import { render, screen } from '@testing-library/react';
+import AppLayout, { AppLayoutProps } from '../../../lib/components/app-layout';
+import { useDynamicOverlap } from '../../../lib/components/internal/hooks/use-dynamic-overlap';
+import { useAppLayoutInternals } from '../../../lib/components/app-layout/visual-refresh/context';
+
+jest.mock('../../../lib/components/internal/hooks/use-visual-mode', () => ({
+  ...jest.requireActual('../../../lib/components/internal/hooks/use-visual-mode'),
+  useVisualRefresh: jest.fn().mockReturnValue(true),
+}));
+
+let positiveHeight = true;
+
+jest.mock('../../../lib/components/internal/hooks/container-queries/utils', () => ({
+  ...jest.requireActual('../../../lib/components/internal/hooks/container-queries/utils'),
+  convertResizeObserverEntry: () => ({ contentBoxHeight: positiveHeight ? 800 : 0 }),
+}));
+
+describe('Dynamic overlap', () => {
+  function ComponentWithDynamicOverlap() {
+    const ref = useDynamicOverlap();
+    const { isDynamicOverlapSet, isDynamicOverlapDisabled } = useAppLayoutInternals();
+    return (
+      <>
+        <div ref={ref} />
+        <div data-testid="is-dynamic-overlap-set">{isDynamicOverlapSet.toString()}</div>
+        <div data-testid="is-dynamic-overlap-disabled">{isDynamicOverlapDisabled.toString()}</div>
+      </>
+    );
+  }
+
+  function ComponentWithoutDynamicOverlap() {
+    const { isDynamicOverlapSet, isDynamicOverlapDisabled } = useAppLayoutInternals();
+    return (
+      <>
+        <div data-testid="is-dynamic-overlap-set">{isDynamicOverlapSet.toString()}</div>
+        <div data-testid="is-dynamic-overlap-disabled">{isDynamicOverlapDisabled.toString()}</div>
+      </>
+    );
+  }
+
+  function renderApp(appLayoutProps?: AppLayoutProps) {
+    render(<AppLayout {...appLayoutProps} />);
+    return {
+      isDynamicOverlapSet: () => screen.getByTestId('is-dynamic-overlap-set').textContent,
+      isDynamicOverlapDisabled: () => screen.getByTestId('is-dynamic-overlap-disabled').textContent,
+    };
+  }
+
+  beforeEach(() => {
+    positiveHeight = true;
+  });
+
+  test('sets dynamic overlap when content header is present', () => {
+    const { isDynamicOverlapSet, isDynamicOverlapDisabled } = renderApp({
+      content: <ComponentWithDynamicOverlap />,
+      contentHeader: 'Content header',
+    });
+    expect(isDynamicOverlapSet()).toBe('true');
+    expect(isDynamicOverlapDisabled()).toBe('false');
+  });
+
+  test('sets dynamic overlap when height is higher than 0', () => {
```

Comment:
Could we have a test case where the height dynamically changes (for example, from `800` to `0`) and we check for the expected values of `isDynamicOverlapSet` and `isDynamicOverlapDisabled`?

Or is this not something that can happen?

Reply:
It's not impossible that these state variables change during the lifecycle of an application but I think it should be rather rare. Some of these events should happen:

1. The height of a child component that uses the dynamic overlap hook changes from 0 to non-0, or the other way around from non-0 to 0
2. A child component that uses the dynamic overlap hook is added or removed
3. `contentHeader` is passed and then not passed anymore, or the other way around
4. The value of the `disableDynamicOverlap` property passed to the app layout changes

I just added a test that checks these state variables after re-rendering, based on the case 2 above. Case 1 would be tricky to test because we would need to simulate a call to the resize observer hook.

Reply:
Alright, that makes sense!